pkg/querier/tripperware/instantquery/instant_query.go

```diff
@@ -0,0 +1,442 @@
+package instantquery
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+	"google.golang.org/grpc/status"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/querier/tripperware"
+	"github.com/cortexproject/cortex/pkg/querier/tripperware/queryrange"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+var (
+	InstantQueryCodec tripperware.Codec = instantQueryCodec{}
+	json                                = jsoniter.Config{
+		EscapeHTML:             false, // No HTML in our responses.
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+	}.Froze()
+)
+
+type PrometheusRequest struct {
+	tripperware.Request
+	Time    int64
+	Stats   string
+	Query   string
+	Path    string
+	Headers http.Header
+}
+
+// GetTime returns time in milliseconds.
+func (r *PrometheusRequest) GetTime() int64 {
+	return r.Time
+}
+
+// GetStart returns always 0 for instant query.
+func (r *PrometheusRequest) GetStart() int64 {
+	return 0
+}
+
+// GetEnd returns always 0 for instant query.
+func (r *PrometheusRequest) GetEnd() int64 {
+	return 0
+}
+
+// GetStep returns always 0 for instant query.
+func (r *PrometheusRequest) GetStep() int64 {
+	return 0
+}
+
+// GetQuery returns the query of the request.
+func (r *PrometheusRequest) GetQuery() string {
+	return r.Query
+}
+
+// WithStartEnd clone the current request with different start and end timestamp.
+func (r *PrometheusRequest) WithStartEnd(int64, int64) tripperware.Request {
+	return r
+}
+
+// WithQuery clone the current request with a different query.
+func (r *PrometheusRequest) WithQuery(query string) tripperware.Request {
+	q := *r
+	q.Query = query
+	return &q
+}
+
+// LogToSpan writes information about this request to an OpenTracing span
+func (r *PrometheusRequest) LogToSpan(sp opentracing.Span) {
+	sp.LogFields(
+		otlog.String("query", r.GetQuery()),
+	)
+}
+
+// GetStats returns the stats of the request.
+func (r *PrometheusRequest) GetStats() string {
+	return r.Stats
+}
+
+// WithStats clones the current `PrometheusRequest` with a new stats.
+func (r *PrometheusRequest) WithStats(stats string) tripperware.Request {
+	q := *r
+	q.Stats = stats
+	return &q
+}
+
+type instantQueryCodec struct {
+	tripperware.Codec
+}
+
+func (resp *PrometheusInstantQueryResponse) HTTPHeaders() map[string][]string {
+	if resp != nil && resp.GetHeaders() != nil {
+		r := map[string][]string{}
+		for _, header := range resp.GetHeaders() {
+			if header != nil {
+				r[header.Name] = header.Values
+			}
```

Comment:
We are overwriting the headers rather than merging the header values with the same key? Looks like that's the original implementation, tbh I am not sure whether this is correct...

Reply:
I think its because `GetHeaders` should never return duplicate keys.. how would this happen, right?

Reply:
Makes sense.